### DISCUSSION TITLE
Using IOrchardHelper in Blazor application

### DISCUSSION THREAD

**Discussion by agriffard** [OrchardCMS/OrchardCore#13611] [General]
I am trying to find out if their is a chance to use IOrchardHelper in a .razor page of a Blazor Server app.

I created a simple test app (https://github.com/agriffard/Zord), referenced the OC CMS targets package and added `builder.Services.AddOrchardCms();` in Program.cs.

I could not add `app.UseOrchardCore();` because it was leading to an `ArgumentException: Value cannot be null or empty. (Parameter 'areaName')` in ModularTenantRouterMiddleware so I commented it.

At this point, the app is not redirected to /setup OC page on loading or responding to this url.

When I add a Page.razor with a directive `@page "/page/{alias}"`, inject a `IOrchardHelper` and use it to load a content by its alias, there is a `NullReferenceException: Object reference not set to an instance of an object.` in _Host.cshtml on:
`<component type="typeof(App)" render-mode="ServerPrerendered" />.`

Is this a dead end to think that it is possible to use OC in a Blazor application?

**Comment by ns8482e** [comment]
@agriffard  I guess you can't reach to `/setup` because you have not added `UseOrchardCore`.

1. Just calling `AddOrchardCms()` and `UseOrchardCore()`  is enough You don't need to call `AddRazorPages()` or `UseRouting()` explicitly.
2. Register Blazor and any other services using orchard specific configure/configureservices extentions.
3. In orchard, app is also an area, so instead of `MapFallbackToPage` use `MapFallbackToAreaPage`.
4. Remove `UseStaticFiles`  from host  pipeline and add in orchard's configure pipeline.

```cs
builder.Services.AddOrchardCms()
.ConfigureServices(svc=> {
    svc.AddServerSideBlazor();
    svc.AddSingleton<WeatherForecastService>();
})
.Configure((a,b,c)=>{
    b.MapBlazorHub();
    a.UseStaticFiles();
    b.MapFallbackToAreaPage("/_Host", "Zord.App");
});
```

**Comment by agriffard** [comment]
@ns8482e  I made the change you suggested but this error is returned:

`InvalidOperationException: Cannot find the fallback endpoint specified by route values: { page: /_Host, area: Zord.App }.`

**Comment by ns8482e** [comment]
I believe you don't have OC's default tenant setup yet that's why you are getting this error

**Comment by ns8482e** [comment]
FallBack to  `MapFallbackToAreaPage("/_Host", "Zord.App")'` when default site is in running state

**Comment by agriffard** [comment]
> FallBack to `MapFallbackToAreaPage("/_Host", "Zord.App")'` when default site is in running state

Thank you @ns8482e. I commented `MapFallbackToAreaPage("/_Host", "Zord.App")'`, run setup and it worked.

**Comment by agriffard** [comment]
I defined a `Page` content type with Alias, Title and HtmlBody parts.
A `Page.razor` with a `@page "/page/{alias}"` directive allows to load a content by this alias.

When I create two pages and want to switch between the corresponding two links that I added in the NavMenu, the content is not reloaded (it works if I click on another Razor page and go to the second page).

Not sure how I can force a page to reload.

**Comment by ns8482e** [comment]
Blazer server application is SPA application. Only part of app is server rendered via signalr

**Comment by ns8482e** [comment]
@agriffard I guess I understood your question now, just looked at your code use parameter set method instead of initialize to see page content changes. Blazor doesn't recreate component and initialize only gets called once.

**Comment by agriffard** [comment]
Thank you @ns8482e. I also figured it out in the mean time, by using the `OnParametersSet` method.

**Comment by hishamco** [comment]
Maybe I came late, but `OnParameterSet` will be triggered when you set a Blazor parameter. It's useful when you need to execute some logic on load, the below is an example I did to load the languages to be displayed in the `LanguageSwitcher` component

https://github.com/oqtane/oqtane.framework/blob/ef27a0d6b0ffab645a610e68890725bba2965059/Oqtane.Client/Themes/Controls/Theme/LanguageSwitcher.razor#L27-L31
